### ISSUE TITLE
Rails 2.3

### ISSUE THREAD

**Issue by merhard** [jswanner/RailsDiff#8]
I love the site. Any chance of getting a diff for rails 2.3.18? I have a bunch of apps to upgrade and this would help a lot.

**Comment by jswanner** [comment]
@merhard thanks for the question.

The process of generating the diffs is completely automated, really generating the whole site is automated. But, when I was first building it all, the process of generating apps for Rails 2.x was completely different than that of 3.x (and 4.x now for that matter), to the point where it would have been difficult to automate the generation of 2.x applications. Which is part of the reason those versions are not included. I just did some testing, and I think I've found a way for it to be done, at least with 2.3.18.

But, aside from the technical issues of generating apps, which are needed for making the diffs. The other reason I didn't include versions below 3.0.0 was that the process to upgrade is much more than just some configuration changes. And, because of that, there are a number of resources and tools available to assist with that process. Things like: https://github.com/rails/rails_upgrade, http://railscasts.com/episodes/225-upgrading-to-rails-3-part-1, http://nerds.airbnb.com/upgrading-airbnb-from-rails-23-to-rails-30/.

There are many things to change in the app that a diff of the generated files will not reveal to you. I can certainly agree with the argument to have diffs for 2.3.17 to 2.3.18, but not as much for 2.3.18 to 4.1.0.

If you just want to see a diff of 2.3.18 to 3.0.0, I have it in a gist: https://gist.github.com/jswanner/11386291.

In summary: I'm hesitant to add versions below 3.0.0, but I'm open to arguments that tell me I'm wrong.

**Comment by jswanner** [comment]
Another hesitation is that this repository already contains a large number of files, and takes up a good amount of disk space. Including all the versions of Rails below 3.0.0 would make the problem exponentially worse.

For some numbers: there are currently 1,892 diffs in the site right now. That's 1,653 diffs comparing all non-prerelease versions of Rails plus some prerelease versions greater than or equal 3.0.0. If I had a diff for every non-prerelease version of Rails that would be 8,001 diffs, and every release of Rails, including prerelease versions, would be 18,915 diffs.

And, for each of these diffs, there are actually 2 HTML files, 1 with limited context lines and another with full file diffs.

_edit: added "greater than or equal 3.0.0" for clarification._

**Comment by merhard** [comment]
@jswanner,
I modified your [Rakefile](https://gist.github.com/merhard/f02db51f21fab55158a4) to allow Rails 2.3 and up.

The number of files is concerning. Perhaps some sort of limit on what can be compared? Say, v2.3 can only be compared up to v3.0, v3.0 to v3.1, and so on? I understand comparing 2.3.18 to 4.1.0 is pretty pointless, but how useful then is 3.0.0 to 4.1.0?

**Comment by jswanner** [comment]
> how useful then is 3.0.0 to 4.1.0?

Fair point. I think the difference is that I might actually upgrade a 3.0.0 app directly to 4.1.0, while I wouldn't do that with 2.3.x, I would upgrade it to 3.0.0 first. Because, I believe the upgrade from 2.3 to 3.x was a big jump, much more so than 3.x to 4.x

> Perhaps some sort of limit on what can be compared? Say, v2.3 can only be compared up to v3.0, v3.0 to v3.1, and so on?

Yeah, I thought about this too. I don't like that the UI won't reflect that limitation, unless a lot is changed on that side to support it; the Rakefile will also need to be changed to add in a concept of a "max version" for any given version.

**Comment by merhard** [comment]
I updated the [Rakefile](https://gist.github.com/merhard/f02db51f21fab55158a4) with a working version (limits 2.3 to be compared through 3.0), though it doesn't help the UI...

**Comment by merhard** [comment]
@jswanner

[A gist showing my changes to the Rakefile and app.js as well as the modified versions.json.](https://gist.github.com/merhard/16954bc45d718f964927)

The Rakefile has a concept of "max version" via the '#max_comparable_tag' method (currently limiting Rails v2.3 to v3.0 and all other versions to v4.2 (effectively no limit since v4.2 doesn't exist)).

The UI reflects the limitation of available diffs. The switch button functionality breaks, but the function is no longer necessary with the limitations.

Thoughts?

**Comment by jswanner** [comment]
@merhard 

Thanks for all your work on this, I'll incorporate your changes in tonight.

> The Rakefile has a concept of "max version" via the '#max_comparable_tag' method (currently limiting Rails v2.3 to v3.0 and all other versions to v4.2 (effectively no limit since v4.2 doesn't exist)).

I'll need to make some changes to this, as I'd prefer no upper bound, until Rails releases a version that makes upgrading difficult, like 3.0 did.

> The switch button functionality breaks, but the function is no longer necessary with the limitations.

Hmm, this is unfortunate, as it does appear people use that. But, you are right, if the UI doesn't let you pick bad values, then switching them to perhaps getting correct ones is not as needed.

**Comment by merhard** [comment]
I created a pull request with my changes, including changing 'v4.2' with 'w' for a max version ('w' > 'v')

https://github.com/jswanner/RailsDiff/pull/9

**Comment by merhard** [comment]
Now that those commits are merged in, I'll close the issue. Thanks
